packages/gatsby/cache-dir/ensure-resources.js

```diff
@@ -5,19 +5,55 @@ import shallowCompare from "shallow-compare"
 
 let isInitialRender = true
 
-// Pass pathname in as prop.
-// component will try fetching resources. If they exist,
-// will just render, else will render null.
-// It will also wait for pageResources
-// before propagating location change to children.
+// ensure-resources.js is the top-level component called by the
+// Router. It is therefore rendered on every page navigation. It's job
+// is to make sure that we have all the resources required to render
+// the page for the props.location. And to fall back to a 404 page if
+// necessary. Once these resources are ensured, then they are passed
+// to props.children, which is actually a function that takes those
+// resources and returns React components (see production-app.js).
+//
+// On the initial render, production-app.js will have already called
+// loader.loadPage. If no resources were found, then ensure-resources
+// will throw an exception, which will end the running js and result
+// in only the pure HTML of the page being displayed.
+//
+// ## Happy Path:
+//
+// When a navigation occurs, navigate.js will call
+// loader.loadPage. `getDerivedStateFromProps` will retrieve that data
+// and return it as props. `shouldComponentUpdate` will check if those
+// resources have changed, and if so, will call render.
+//
+// ## Page data or component blocked:
+//
+// This can happen for e.g if an ad blocker blocks json or js, or
+// internet connection cutoff halfway through request. In this case,
+// when navigation occurs, navigate.js calls loader.loadPage, which
+// returns null. `getDerivedStateFromProps` updates props with null
+// pageResources. `shouldComponentUpdate` sees no resources, so calls
+// `retryResources`, which performs a request for the page html. If
+// the html page exists, then we have a fallback, so we reload the
+// entire page, resulting in the pure HTML being rendered.
```

Comment:
This is super cool — extra security for Gatsby sites

Reply:
Just to be clear, I didn't introduce this functionality. I merely refactored and documented it 👼